### PR TITLE
OpenCS: Add brush settings to terrain edit

### PR DESCRIPTION
**Pull request by unelsson**
(clean) rebase of https://github.com/OpenMW/openmw/pull/1665
- Brush settings window for terrain texture editing mode

### CODE REVIEW

File: apps/opencs/view/render/terraintexturemode.cpp [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+{
+    mBrushTextureLabel = "Brush: " + mBrushTexture;
+    selectedBrush = new QLabel(QString::fromUtf8(mBrushTextureLabel.c_str()), this);
+
+    const std::string& iconPoint = ":scenetoolbar/brush-point";
+    const std::string& iconSquare = ":scenetoolbar/brush-square";
+    const std::string& iconCircle = ":scenetoolbar/brush-circle";
+    const std::string& iconCustom = ":scenetoolbar/brush-custom";
+
+    TextureBrushButton *buttonPoint = new TextureBrushButton(QIcon (QPixmap (iconPoint.c_str())), "", this);
```
Comment (zinnschlag) [inline apps/opencs/view/render/terraintexturemode.cpp]: That is not how you convert between std::string and QString (at least not in Qt4). Messes up the encoding. There is an fromUtf8 function.

File: apps/opencs/view/render/terraintexturemode.hpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        public:
+            BrushSizeControls(const QString &title, QWidget *parent);
+
+        private:
+            QSlider *brushSizeSlider;
+            QSpinBox *brushSizeSpinBox;
+            QHBoxLayout *layoutSliderSize;
+    };
+
+    class TextureBrushButton : public QPushButton
```
Comment (zinnschlag) [inline apps/opencs/view/render/terraintexturemode.hpp]: Is there are reason why you are using a custom class for this?

**Comment by zinnschlag** [comment]
Did another review (final test run still missing). I made some inline comments. Seems mostly okay. But I noticed that you don't handle drops of textures from the texture resource table.

**Comment by unelsson** [comment]
How would the intended behavior for dropping texture from resource table be? Creating a new index to landtextures-table that points to the texture filename? There's a possible problem with that regarding Terraintexture primary edit #1680 branch - if user uses touchlandcommand after using a texture from resource table, it might not handle overlapping indexes properly.

File: apps/opencs/view/render/terraintexturemode.cpp (hunk 1 above)
Comment (unelsson) [reply apps/opencs/view/render/terraintexturemode.cpp]: Hm. Okay. I changed it to QString::fromStdString as QString::fromUtf8 takes only const char *str.

**Comment by zinnschlag** [comment]
Finally got around to test this. Does not work for me. The texture mode button is black and no other buttons (or other GUI elements) are shown.

**Comment by unelsson** [comment]
My guess is that mode button is black because there is no icon, I didn't include icons here, as there's a separate [https://github.com/OpenMW/openmw/pull/1677](pull request) that modifies same files quite extensively. Brush button window opens by pressing MMB.

**Comment by unelsson** [comment]
No need to keep this open, as development of https://github.com/OpenMW/openmw/pull/1680 is ongoing and it has everything this has.

**Comment by zinnschlag** [comment]
Yeah, the whole GUI is button based. Clicking in the scene should not open any GUI.

**Pull request closed without merging**